src/fsverity/hashvalue.rs

```diff
@@ -1,27 +1,121 @@
+use core::{fmt, hash::Hash};
+
+use hex::FromHexError;
 use sha2::{digest::FixedOutputReset, digest::Output, Digest, Sha256, Sha512};
+use zerocopy::{FromBytes, Immutable, IntoBytes, KnownLayout, Unaligned};
 
 pub trait FsVerityHashValue
 where
-    Self: Eq + AsRef<[u8]> + Clone,
+    Self: Clone,
     Self: From<Output<Self::Digest>>,
+    Self: FromBytes + Immutable + IntoBytes + KnownLayout + Unaligned,
+    Self: Hash + Eq,
+    Self: fmt::Debug,
 {
-    type Digest: Digest + FixedOutputReset + std::fmt::Debug;
+    type Digest: Digest + FixedOutputReset + fmt::Debug;
     const ALGORITHM: u8;
     const EMPTY: Self;
+    const ID: &str;
+
+    fn from_hex(hex: impl AsRef<[u8]>) -> Result<Self, FromHexError> {
+        let mut value = Self::EMPTY;
+        hex::decode_to_slice(hex.as_ref(), value.as_mut_bytes())?;
+        Ok(value)
+    }
+
+    fn from_object_dir_and_basename(
+        dirnum: u8,
+        basename: impl AsRef<[u8]>,
+    ) -> Result<Self, FromHexError> {
+        let expected_size = 2 * (size_of::<Self>() - 1);
+        let bytes = basename.as_ref();
+        if bytes.len() != expected_size {
+            return Err(FromHexError::InvalidStringLength);
+        }
+        let mut result = Self::EMPTY;
+        result.as_mut_bytes()[0] = dirnum;
+        hex::decode_to_slice(bytes, &mut result.as_mut_bytes()[1..])?;
+        Ok(result)
+    }
+
+    fn from_object_pathname(pathname: impl AsRef<[u8]>) -> Result<Self, FromHexError> {
```

Comment:
Not new but this one looks unit testable.

Also here and elsewhere it's more efficient to do the "helper function" pattern where there's a generic wrapper for a concrete function that takes a concrete `&[u8]` does the real work.

Reply:
Adding unit tests found an actual bug, thanks!